compiler/rustc_expand/src/derive_macro_expansion.rs

```diff
@@ -0,0 +1,125 @@
+use std::cell::Cell;
+use std::ptr;
+
+use rustc_ast::tokenstream::TokenStream;
+use rustc_data_structures::svh::Svh;
+use rustc_middle::ty::TyCtxt;
+use rustc_span::profiling::SpannedEventArgRecorder;
+use rustc_span::LocalExpnId;
+
+use crate::base::ExtCtxt;
+use crate::errors;
+
+pub(super) fn provide_derive_macro_expansion<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    key: (LocalExpnId, Svh, &'tcx TokenStream),
+) -> Result<&'tcx TokenStream, ()> {
+    let (invoc_id, _macro_crate_hash, input) = key;
+
+    let res = with_context(|(ecx, client)| {
+        let span = invoc_id.expn_data().call_site;
+        let _timer = ecx.sess.prof.generic_activity_with_arg_recorder(
+            "expand_derive_proc_macro_inner",
+            |recorder| {
+                recorder.record_arg_with_span(ecx.sess.source_map(), ecx.expansion_descr(), span);
+            },
+        );
+        let proc_macro_backtrace = ecx.ecfg.proc_macro_backtrace;
+        let strategy = crate::proc_macro::exec_strategy(ecx);
+        let server = crate::proc_macro_server::Rustc::new(ecx);
+        let res = match client.run(&strategy, server, input.clone(), proc_macro_backtrace) {
+            // FIXME(pr-time): without flattened some (weird) tests fail, but no idea if it's correct/enough
+            Ok(stream) => Ok(tcx.arena.alloc(stream.flattened()) as &TokenStream),
+            Err(e) => {
+                ecx.dcx().emit_err({
```

Comment:
Anything that can use `tcx` instead of `ecx` should use `tcx`.
`sess` is available from `tcx` in particular.

Reply:
changed to `tcx.dcx()...`